File: 78-subsets/subsets.java
import java.util.*;

class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        List<Integer> subset = new ArrayList<>();
        backtrack(nums, 0, subset, result);
        return result;
    }
    
    private void backtrack(int[] nums, int start, List<Integer> subset, List<List<Integer>> result) {
        result.add(new ArrayList<>(subset)); // Add the current subset to the result
        
        // Generate subsets by including each element one by one
        for (int i = start; i < nums.length; i++) {
            subset.add(nums[i]); // Include the current element
            backtrack(nums, i + 1, subset, result); // Recursively generate subsets
            subset.remove(subset.size() - 1); // Backtrack: remove the last element to backtrack
        }
    }
}

File: 78-subsets/subsets.cpp
class Solution {
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int>> result;
        vector<int> current;
        backtrack(nums, 0, current, result);
        return result;
    }
    
private:
    void backtrack(vector<int>& nums, int start, vector<int>& current, vector<vector<int>>& result) {
        result.push_back(current);  // Add the current subset to the result
        for (int i = start; i < nums.size(); ++i) {
            current.push_back(nums[i]);  // Include the current element
            backtrack(nums, i + 1, current, result);  // Recur with the next element
            current.pop_back();  // Exclude the current element (backtrack)
        }
    }
};